lib/fastlane/plugin/firebase_app_distribution/client/firebase_app_distribution_api_client.rb

```diff
@@ -123,57 +131,56 @@ def upload_binary(app_id, binary_path, platform)
       # Takes at least POLLING_INTERVAL_SECONDS between polling get_upload_status
       #
       # args
-      #   project_number - Firebase project number
-      #   app_id - Firebase app ID
+      #   app_name - Firebase App resource name
       #   binary_path - Absolute path to your app's aab/apk/ipa file
       #
-      # Returns the release_id of the uploaded release.
+      # Returns the release_name of the uploaded release.
       #
       # Crashes if the number of polling retries exceeds MAX_POLLING_RETRIES or if the binary cannot
       # be uploaded.
-      def upload(project_number, app_id, binary_path, platform)
+      def upload(app_name, binary_path, platform)
         binary_type = binary_type_from_path(binary_path)
 
-        upload_token = get_upload_token(project_number, app_id, binary_path)
-        upload_status_response = get_upload_status(app_id, upload_token)
-        if upload_status_response.success? || upload_status_response.already_uploaded?
-          UI.success("✅ This #{binary_type} has been uploaded before. Skipping upload step.")
-        else
-          unless upload_status_response.in_progress?
-            UI.message("⌛ Uploading the #{binary_type}.")
-            upload_binary(app_id, binary_path, platform)
-          end
-          MAX_POLLING_RETRIES.times do
-            upload_status_response = get_upload_status(app_id, upload_token)
-            if upload_status_response.success? || upload_status_response.already_uploaded?
-              UI.success("✅ Uploaded the #{binary_type}.")
-              break
-            elsif upload_status_response.in_progress?
-              sleep(POLLING_INTERVAL_SECONDS)
+        UI.message("⌛ Uploading the #{binary_type}.")
+        operation_name = upload_binary(app_name, binary_path, platform)
+
+        upload_status_response = get_upload_status(operation_name)
+        MAX_POLLING_RETRIES.times do
+          if upload_status_response.success?
+            UI.success("✅ Uploaded the #{binary_type} and created release #{upload_status_response.release_version}.")
+            break
+          elsif upload_status_response.release_updated?
+            UI.success("✅ The same #{binary_type} was found, and the existing #{upload_status_response.release_version} release has been updated.")
```

Comment:
"same"might be misleading, since the provisioning profile changed. I don't know how explicit we want to be, but we could say something along the lines "✅ Uploaded #{binary_type} successfully; updated provisioning profile of existing release {upload_status_response.release_version}"

I guess we should agree on verbiage for all three CLI tools and make it identical (or at least very similar)

Reply:
This won't be the case for android. This can happen if they add testers to an existing release, for example.

Reply:
No, not really. On Android there's always exactly one binary per release. So either we get `RELEASE_CREATED` or `RELEASE_UNMODIFIED`. The `RELEASE_UPDATED` can only happen on iOS when the "instance identifier" (http://google3/java/com/google/firebase/appdistro/boq/appmetadataextractor/service/ios/IpaMetadata.java?l=63-73&rcl=390029034) remains the same, but the sha-256 of the entire binary changes.

For Android the instance identifier is the sha-1 of the entire binary. So if the binary changes, both the sha-1 and sha-256 change and a new release is created. If the binary doesn't change, it's a duplicate upload.

Reply:
Sorry, I know about the iOS provisioning profile case. I just realized that what I was referring to is unrelated to upload and more of an enable access concern.

Reply:
Updated to use your wording instead for this (I dig it!)